common/retry_delays.go

```diff
@@ -0,0 +1,134 @@
+// If maxRetries is set to 0, it means infinite retries are allowed.
+//
+// You can also abort the retrying wrapping the error ErrAbort into the result
+// this is useful if there are some conditions that should not be retried
+type RetryDelays struct {
+	Delays []types.Duration `mapstructure:"Delays"`
+	// MaxAttempts is the maximum number of retries to attempt.
+	// if MaxAttempts is 0, it means infinite retries.
+	// if MaxAttempts is 1, it means no retries will be attempted.
+	MaxAttempts int `mapstructure:"MaxAttempts"`
+}
+
+func (r *RetryDelays) Validate() error {
+	if r == nil {
+		return fmt.Errorf("retry delays cannot be nil. %w", ErrInvalidConfig)
+	}
+	if len(r.Delays) == 0 {
+		return fmt.Errorf("retry delays cannot be empty. %w", ErrInvalidConfig)
+	}
+	for _, delay := range r.Delays {
+		if delay.Duration <= 0 {
+			return fmt.Errorf("retry delay must be greater than zero, got %s. %w",
+				delay.Duration, ErrInvalidConfig)
+		}
+	}
+	if r.MaxAttempts < 0 {
+		return fmt.Errorf("max retries cannot be negative, got %d. %w",
+			r.MaxAttempts, ErrInvalidConfig)
+	}
+	return nil
+}
+
+func (r *RetryDelays) String() string {
+	if r == nil {
+		return "RetryDelays is nil"
+	}
+	return fmt.Sprintf("RetryDelays{Delays: %v, MaxRetries: %d}", r.Delays, r.MaxAttempts)
+}
+
+func (r *RetryDelays) InfiniteRetries() bool {
+	// Infinite retries are allowed if MaxAttempts is 0.
+	return r.MaxAttempts == 0
+}
+
+func (r *RetryDelays) Delay(attempt int) time.Duration {
+	if r == nil || len(r.Delays) == 0 {
+		return 0
+	}
+	return r.Delays[min(attempt, len(r.Delays)-1)].Duration
+}
+
+// Execute executes the provided function with retry logic.
+// retryDelaysConfig: it's the RetryDelays struct that holds the retry delays and the maximum number of retries.
+// ctx: the context to use for cancellation.
+// logFunc: a function to log messages, if nil a silent logger will be used.
+// name: the name of the operation, used for logging.
+// fn: the function to execute, it should return a result of type T and an error.
+// If the function returns an error that is wrapped with ErrAbort,
+// the execution will be aborted and no more retries will be attempted.
+func Execute[T any](retryDelaysConfig *RetryDelays,
+	ctx context.Context,
+	logFunc func(format string, args ...interface{}),
+	name string,
+	payloadFunc func() (T, error)) (T, error) {
+	var zero T
+	if err := retryDelaysConfig.Validate(); err != nil {
+		return zero, err
+	}
+	if logFunc == nil {
+		// if logger is nil, we create a silent logger
+		logFunc = func(format string, args ...interface{}) {}
+	}
+
+	attempt := 0
+	for attempt := 0; retryDelaysConfig.InfiniteRetries() || attempt < retryDelaysConfig.MaxAttempts; attempt++ {
+		delay := retryDelaysConfig.Delay(attempt)
+		logFunc("executing %s try %d/%d (next delay: %s)",
+			name, attempt+1, retryDelaysConfig.MaxAttempts, delay.String())
```

Comment:
Should we tweak the log message in case we have infinite retries? In particular, it probably does not make sense to output `%d/%d` for the attempts, but maybe just output the current attempt. WDYT?